pkg/controller/packageregistry/controller.go

```diff
@@ -0,0 +1,378 @@
+		CACertRotation:        r.CACertRotation,
+		CertRotation:          r.CertRotation,
+		GarbageCollectSecrets: true,
+	}.ReconcileCAAndHTTPCerts(ctx)
+	if results.HasError() {
+		_, err := results.Aggregate()
+		k8s.MaybeEmitErrorEvent(r.recorder, err, &epr, events.EventReconciliationError, "Certificate reconciliation error: %v", err)
+		return results, status
+	}
+
+	configSecret, err := reconcileConfig(ctx, r, epr, meta)
+	if err != nil {
+		return results.WithError(err), status
+	}
+
+	// build a hash of various inputs to rotate Pods on any change
+	configHash, err := buildConfigHash(r.K8sClient(), epr, configSecret)
+	if err != nil {
+		return results.WithError(fmt.Errorf("build config hash: %w", err)), status
+	}
+
+	deploy, err := r.reconcileDeployment(ctx, epr, configHash, meta)
+	if err != nil {
+		return results.WithError(fmt.Errorf("reconcile deployment: %w", err)), status
+	}
+
+	status, err = r.getStatus(ctx, epr, deploy)
+	if err != nil {
+		return results.WithError(fmt.Errorf("calculating status: %w", err)), status
+	}
+
+	return results, status
+}
+
+func newStatus(epr eprv1alpha1.PackageRegistry) eprv1alpha1.PackageRegistryStatus {
+	status := epr.Status
+	status.ObservedGeneration = epr.Generation
+	return status
+}
+
+func (r *ReconcilePackageRegistry) validate(ctx context.Context, epr eprv1alpha1.PackageRegistry) error {
+	span, vctx := apm.StartSpan(ctx, "validate", tracing.SpanTypeApp)
+	defer span.End()
+
+	if _, err := epr.ValidateCreate(); err != nil {
+		ulog.FromContext(ctx).Error(err, "Validation failed")
+		k8s.MaybeEmitErrorEvent(r.recorder, err, &epr, events.EventReasonValidation, err.Error())
+		return tracing.CaptureError(vctx, err)
+	}
+
+	return nil
+}
+
+func NewService(epr eprv1alpha1.PackageRegistry, meta metadata.Metadata) *corev1.Service {
+	svc := corev1.Service{
+		ObjectMeta: epr.Spec.HTTP.Service.ObjectMeta,
+		Spec:       epr.Spec.HTTP.Service.Spec,
+	}
+
+	svc.ObjectMeta.Namespace = epr.Namespace
+	svc.ObjectMeta.Name = HTTPServiceName(epr.Name)
+
+	selector := epr.GetIdentityLabels()
+	ports := []corev1.ServicePort{
+		{
+			Name:     epr.Spec.HTTP.Protocol(),
+			Protocol: corev1.ProtocolTCP,
+			Port:     HTTPPort,
+		},
+	}
+
+	return defaults.SetServiceDefaults(&svc, meta, selector, ports)
+}
+
+func buildConfigHash(c k8s.Client, epr eprv1alpha1.PackageRegistry, configSecret corev1.Secret) (string, error) {
+	// build a hash of various settings to rotate the Pod on any change
+	configHash := fnv.New32a()
+
+	// - in the Elastic Package Registry configuration file content
+	_, _ = configHash.Write(configSecret.Data[ConfigFilename])
+
+	// - in the Elastic Package Registry TLS certificates
+	if epr.Spec.HTTP.TLS.Enabled() {
+		var tlsCertSecret corev1.Secret
+		tlsSecretKey := types.NamespacedName{Namespace: epr.Namespace, Name: certificates.InternalCertsSecretName(eprv1alpha1.Namer, epr.Name)}
+		if err := c.Get(context.Background(), tlsSecretKey, &tlsCertSecret); err != nil {
```

Comment:
Must not instaniate new context but re-use existing context: 
- APM tracing 
- logger is in context 
- proper cancellation propagation

Reply:
changed here 8f49e4f9246b705c66dd0c90c23311419c31ddd8